src/ubloxGPS.cpp

```diff
@@ -273,100 +275,93 @@ void ubloxGPS::updateGPS(void)
     }
 }
 
-void ubloxGPS::setOn(lib_config_t &config)
+int ubloxGPS::setOn(lib_config_t &config)
 {
     LOCK();
 
-    if(serial)
+    if(isInterfaceUart())
     {
         serial->begin((uint32_t)UBX_BAUDRATE_DEFAULT);
     }
 
     pwr_enable(true);
 
-    gpsStatus = GPS_STATUS_FIXING;
+    NAMED_SCOPE_GUARD(exitScope, {
+        // Turn device off
+        if(log_enabled) Loglib.error("UBX Initialization failed");
+        gpsStatus = GPS_STATUS_ERROR;
+        pwr_enable(false);
+    });
     delay(1000); //delay 1000ms to wait for ubloxGPS get ready for receive UBX command (tested value)
 
-    if(serial)
+    if(isInterfaceUart())
     {
         // set high baudrate to improve performance
-        setBaudrate((ubx_baudrate_t)config.baudrate);
-        if(log_enabled) Loglib.info("UBX baudrate is %lu", config.baudrate);
-    }
-
-    if(tx_ready_mcu_pin != PIN_INVALID && tx_ready_gps_pin != PIN_INVALID)
-    {
+        CHECK_TRUE(setBaudrate((ubx_baudrate_t)config.baudrate), SYSTEM_ERROR_IO);
+        if(log_enabled)
+            Loglib.info("UBX baudrate is %lu", config.baudrate);
+    } else if (isInterfaceSpi()) {
+        CHECK_FALSE((tx_ready_mcu_pin == PIN_INVALID) || (tx_ready_gps_pin == PIN_INVALID), SYSTEM_ERROR_INVALID_ARGUMENT);
+        // TODO: break this conditional up to exit on queue create error
         if(!tx_ready_queue && !os_queue_create(&tx_ready_queue, sizeof(uint8_t), 1, nullptr))
         {
             pinMode(tx_ready_mcu_pin, INPUT);
             if(!attachInterrupt(tx_ready_mcu_pin, &ubloxGPS::txReadyHandler, this, FALLING))
             {
                 os_queue_destroy(tx_ready_queue, nullptr);
                 tx_ready_queue = nullptr;
+                return SYSTEM_ERROR_NO_MEMORY;
             }
         }
 
-        if(tx_ready_queue)
-        {
-            // first process all bytes off of the port, reconfiguring seems to drop
-            // some startup frames that are useful debug indications
-            processBytes();
+        // first process all bytes off of the port, reconfiguring seems to drop
+        // some startup frames that are useful debug indications
+        processBytes();
```

Comment:
I agree. It would have to be programmed at least once at manufacturing or checked at startup. One concern is that the customer can fork this repo and make their own changes that deviate from the settings here.